strings/strings.1
.Dd December 2008
.Dt STRINGS 1

.Sh NAME

.Nm strings
.Nd find printable strings in files

.Sh SYNOPSIS

.Nm strings
.Ar [-a][-t format][-n number][file...]

.Sh DESCRIPTION

    The strings utility shall look for printable strings in regular files and
shall write those strings to standard output. A printable string is any
sequence of four (by default) or more printable characters terminated by a
<newline> or NUL character. Additional implementation-defined strings may be
written; see localedef.

.Sh OPTIONS

    The strings utility shall conform to the Base Definitions volume of IEEE
Std 1003.1-2001, Section 12.2, Utility Syntax Guidelines.

    The following options shall be supported:

    -a
        Scan files in their entirety. If -a is not specified, it is
implementation-defined what portion of each file is scanned for strings.
    -n  number
        Specify the minimum string length, where the number argument is a
positive decimal integer. The default shall be 4.
    -t  format
        Write each string preceded by its byte offset from the start of the
file. The format shall be dependent on the single character used as the
format option-argument:

        d
            The offset shall be written in decimal.
        o
            The offset shall be written in octal.
        x
            The offset shall be written in hexadecimal.

.Sh OPERANDS

    The following operand shall be supported:

    file
        A pathname of a regular file to be used as input. If no file operand
is specified, the strings utility shall read from the standard input.

.Sh STDIN

    See the INPUT FILES section.

.Sh INPUT FILES

    The input files named by the utility arguments or the standard input
shall be regular files of any format.

.Sh ENVIRONMENT VARIABLES

    The following environment variables shall affect the execution of
strings:

    LANG
        Provide a default value for the internationalization variables that
are unset or null. (See the Base Definitions volume of IEEE Std 1003.1-2001,
Section 8.2, Internationalization Variables for the precedence of
internationalization variables used to determine the values of locale
categories.)
    LC_ALL
        If set to a non-empty string value, override the values of all the
other internationalization variables.
    LC_CTYPE
        Determine the locale for the interpretation of sequences of bytes of
text data as characters (for example, single-byte as opposed to multi-byte
characters in arguments and input files) and to identify printable strings.
    LC_MESSAGES
        Determine the locale that should be used to affect the format and
contents of diagnostic messages written to standard error.
    NLSPATH
        [XSI] [Option Start] Determine the location of message catalogs for
the processing of LC_MESSAGES . [Option End]

.Sh ASYNCHRONOUS EVENTS

    Default.

.Sh STDOUT

    Strings found shall be written to the standard output, one per line.

    When the -t option is not specified, the format of the output shall be:

    "%s", <string>

    With the -t o option, the format of the output shall be:

    "%o %s", <byte offset>, <string>

    With the -t x option, the format of the output shall be:

    "%x %s", <byte offset>, <string>

    With the -t d option, the format of the output shall be:

    "%d %s", <byte offset>, <string>

.Sh STDERR

    The standard error shall be used only for diagnostic messages.

.Sh OUTPUT FILES

    None.

.Sh EXTENDED DESCRIPTION

    None.

.Sh EXIT STATUS

    The following exit values shall be returned:

     0
        Successful completion.
    >0
        An error occurred.

.Sh CONSEQUENCES OF ERRORS

    Default.

The following sections are informative.
.Sh APPLICATION USAGE

    By default the data area (as opposed to the text, "bss", or header areas)
of a binary executable file is scanned. Implementations document which areas
are scanned.

    Some historical implementations do not require NUL or <newline>
terminators for strings to permit those languages that do not use NUL as a
string terminator to have their strings written.

.Sh EXAMPLES

    None.

.Sh RATIONALE

    Apart from rationalizing the option syntax and slight difficulties with
object and executable binary files, strings is specified to match historical
practice closely. The -a and -n options were introduced to replace the
non-conforming - and - number options.

    The -o option historically means different things on different
implementations. Some use it to mean " offset in decimal", while others use
it as " offset in octal". Instead of trying to decide which way would be
least objectionable, the -t option was added. It was originally named -O to
mean "offset", but was changed to -t to be consistent with od.

    The ISO C standard function isprint() is restricted to a domain of
unsigned char. This volume of IEEE Std 1003.1-2001 requires implementations
to write strings as defined by the current locale.

.Sh FUTURE DIRECTIONS

    None.

.Sh SEE ALSO

    localedef, nm




NEWS

Version 5 - unreleased

	- New utilities: cmp, tput, tabs, stty, uname, uniq

	- Bug fixes.

Version 4 - December 9, 2006

	- import codebase into git repository on git.kernel.org
	- fix compiler warnings on newer compilers
	- add x86-64 support
	- add WIP utility: pax
	- add utilities: env, id, strings, time
	- add simple "it works" tests
	- create per-util status files
	- drop hand-rolled 'compress', import BSD 'compress' instead

Version 3 - February 7, 2005

	- Add copyright/license info to each source file.
	- Make sure bug report address points to SourceForge.

	- New utilities: grep, pathchk, pwd, split, tty.
	- New work-in-progress (WIP) utility: mv.

	- Fix many 'compress' bugs.
	- Security fixes.

	- (internal) Move compress engine into lib.

Version 2 - January 26, 2005

	- Update README with project goals/manifesto,
	  and some URLs.

	- New utilities: compress, comm, date, what, wc, renice,
	  ipcrm, kill, rm.

	- Bug fixes.

Version 1 - August 21, 2004

	- Initial release.



strings/strings.cc
/*
 * Copyright 2004-2006 Jeff Garzik <jgarzik@pobox.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.  If not, write to
 * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */


#ifndef HAVE_CONFIG_H
#error missing autoconf-generated config.h.
#endif
#include "posixutils-config.h"

#include <assert.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <ctype.h>
#include <errno.h>
#include <libpu.h>

static const char doc[] =
N_("strings - find printable strings in files");

#define args_doc file_args_doc

static struct argp_option options[] = {
	{ "all", 'a', NULL, 0,
	  N_("Scan files in their entirety") },
	{ "bytes", 'n', "NUMBER", 0,
	  N_("Specify the minimum string length") },
	{ "radix", 't', "FORMAT", 0,
	  N_("Write each string preceded by its byte offset from the start of the file") },
	{ }
};

static int strings_actor(struct walker *w, const char *fn, int fd);
static error_t parse_opt (int key, char *arg, struct argp_state *state);
static const struct argp argp = { options, parse_opt, args_doc, doc };

static struct walker walker;

static unsigned int opt_strlen = 4;
static enum opt_prefix {
	PFX_NONE,
	PFX_DECIMAL,
	PFX_OCTAL,
	PFX_HEX
} opt_prefix = PFX_NONE;

enum random_constants {
	BUFLEN			= 8192,
};

static struct buf_state {
	char	buf[BUFLEN];
	int	used;
	off_t	offset;
} st;

static char *find_sep(char *s, size_t n)
{
	int ch;

	while (n > 0) {
		ch = *s;
		if ((ch == '\n') || (ch == 0))
			return s;

		s++;
		n--;
	}

	return NULL;
}

static int string_print(const char *s)
{
	char ofs[64];

	switch (opt_prefix) {
	case PFX_NONE:
		ofs[0] = 0;
		break;
	case PFX_DECIMAL:
		sprintf(ofs, "%Lu", (unsigned long long) st.offset);
		break;
	case PFX_OCTAL:
		sprintf(ofs, "%Lo", (unsigned long long) st.offset);
		break;
	case PFX_HEX:
		sprintf(ofs, "%Lx", (unsigned long long) st.offset);
		break;
	}

	return printf("%s%s\n", ofs, s);
}

static int strings_buf(void)
{
	char *p, *sep = NULL;
	size_t len, diff;

	p = st.buf;
	len = st.used;
	while (1) {
		unsigned int i;

		sep = find_sep(p, len);
		if (!sep) {
			if (len > (BUFLEN / 2)) {
				diff = len - (BUFLEN / 2);
				p += diff;
				len -= diff;
			}

			assert(p != st.buf);

			memmove(st.buf, p, len);
			st.used = len;
			st.offset += p - st.buf;

			return 0;
		}

		*sep = 0;

		diff = sep - p;
		for (i = 0; i < diff; i++)
			if (!isprint(p[diff - i - 1]))
				break;

		if (i >= opt_strlen)
			if (string_print(p) < (int)opt_strlen)
				return 1;

		p = sep + 1;
		len -= diff + 1;
	}
}

static int strings_actor(struct walker *w, const char *fn, int fd)
{
	ssize_t rrc;

	memset(&st, 0, sizeof(st));

	while (1) {
		rrc = read(fd, st.buf + st.used, BUFLEN - st.used);
		if (rrc < 0) {
			fprintf(stderr, "\n");
			perror(fn);
			return 1;
		}
		if (rrc == 0)
			break;

		st.used += rrc;

		if (strings_buf())
			return 1;
	}

	return 0;
}

static error_t parse_opt (int key, char *arg, struct argp_state *state)
{
	long l;
	char *endp = NULL;

	PU_OPT_BEGIN
	PU_OPT_IGNORE('a')

	case 'n':
		errno = 0;
		l = strtol(arg, &endp, 10);
		if (errno || (l < 1))
			argp_usage(state);
		opt_strlen = l;
		break;
	case 't':
		break;

	PU_OPT_ARG
	PU_OPT_DEFAULT
	PU_OPT_END
}

int main (int argc, char *argv[])
{
	walker.argp			= &argp;
	walker.flags			= WF_NO_FILES_STDIN;
	walker.cmdline_fd		= strings_actor;
	return walk(&walker, argc, argv);
}
